### ISSUE TITLE
Attachment preview shows only blank image

### ISSUE THREAD

**Issue by martgil** [FlowCrypt/flowcrypt-android#2894]
When previewing an encrypted image attachment from an email in the FlowCrypt android app, the image no longer displays and instead replaced by a seemed to be blank image.

<img src="https://github.com/user-attachments/assets/dbbeef89-f064-49a1-b537-0b54444a6a5b" width="45%">

I've tested this issue on both Android 11 (emulator) and 13 (real android device). The attachment download feature continues to work well on the other hand.

Steps to reproduce:

1. Open FlowCrypt Android app version 1.5.9.
2. Open a secure email that contains encrypted image attachment (.jpg file in my testing case).
3. Tap the "preview" icon button and notice that it will then start to decrypt the file and will ask you which application you want to use for opening it where it will suggest the image viewer such as Google Photos. But right there when the Google Photos appear, the image will blank and will not show the preview of the image.

**Comment by DenBond7** [comment]
Thanks, Mart. It should be retested after the `conversation view` will be released.

**Comment by martgil** [comment]
Thanks, @DenBond7! I'll be ready to test that out. I'll try to keep an eye on the release, but if I miss it, please feel free to ping me, and I'll test it out.

**Comment by DenBond7** [comment]
> please feel free to ping me, and I'll test it out.

@martgil Ping )). It can be already tested in https://github.com/FlowCrypt/flowcrypt-android/pull/2821

**Comment by martgil** [comment]
I have tested this out in #2821 and here are the details:

- Testing it from my actual device (Android 13) with Google Photos installed (default image): After I tap the preview attachment and the app starts downloading it, the warning message saying `Sorry, you don't have an application that can open this file` appears at the end of this process.

- But testing it from the android emulator shows different behavior where the images retains blank: 
 
[Screencast from 2024-12-04 11-16-40.webm](https://github.com/user-attachments/assets/2556e207-4fe8-4e55-88b3-dcbd459a8942)

I'd also like to raise this observation regarding attachment download. When downloading the attachment for the first time, the download notification will appear and will show sign of file being downloaded. However, doing it repeatedly won't show the same notification again though the animation of the download within the attachment icons happens but the notification of the download finishes never occur again. Maybe an additional toast can be added so that even through the notification progress bar never pops, at least there will be information from the app that the file has been downloaded. This is just my personal experience and just my opinion with the app regarding file download.

**Comment by DenBond7** [comment]
Thank you @martgil. It looks like I need additional investigation.

> I'd also like to raise this observation regarding attachment download. ... Maybe an additional toast can be added so that even through the notification progress bar never pops, at least there will be information from the app that the file has been downloaded. 

Could you share more details(maybe a video)?

You've tested the following logic:

1. A user clicks on the preview button
2. the app downloads and stores an attachment in the RAM cache(temporary, can be dropped from this cache based on low RAM size)
3. the app opens the cached attachment via existing apps
4. if a user navigates back and clicks on the already cached attachment the app checks if the attachment is still in the cache, gets it(or downloads it again), and displays it via existing apps.

**Comment by martgil** [comment]
Hello @DenBond7, I've shared the video screen recording for the scenarios in attachment preview as well as the downloading process where notification pop up (where notification never gets pops when the user downloads it for the second time but download succeeds).

**Comment by DenBond7** [comment]
Hi @martgil,

1. > When previewing an encrypted image attachment from an email in the FlowCrypt android app, the image no longer displays and instead replaced by a seemed to be blank image.

I'm able to reproduce the issue. But it happens only for the internal Gallery app for AOSP devices like emulators. That's because this app is [deprecated](https://android.googlesource.com/platform/packages/apps/Gallery/+/c7325276f1fb4ae73823361c0ea0dba36b85264a/README) and isn't maintained anymore. I've checked a lot of Android apps(like Google Photos, Google Files, some image viewers, and different file managers on a few devices) and all of them work well.

2. > Testing it from my actual device (Android 13) with Google Photos installed (default image): After I tap the preview attachment and the app starts downloading it, the warning message saying Sorry, you don't have an application that can open this file appears at the end of this process.

I can't reproduce it. If you still have this problem please make a video for me.

3. > However, doing it repeatedly won't show the same notification again though the animation of the download within the attachment icons happens but the notification of the download finishes never occur again. Maybe an additional toast can be added so that even through the notification progress bar never pops, at least there will be information from the app that the file has been downloaded. This is just my personal experience and just my opinion with the app regarding file download.

I've improved this scenario. Please try.

**Comment by martgil** [comment]
Hi @DenBond7 - I'm testing this one now. I'll have more details on this one too. I'll be using the https://github.com/FlowCrypt/flowcrypt-android/pull/2948 to test it and compare it with 1.6.1 play store version. I'll be in touch soon.

**Comment by martgil** [comment]
Hi again @DenBond7,

I'm done testing checking this and here are my observations.

> I'm able to reproduce the issue. But it happens only for the internal Gallery app for AOSP devices like emulators. That's because this app is deprecated and isn't maintained anymore. I've checked a lot of Android apps(like Google Photos, Google Files, some image viewers, and different file managers on a few devices) and all of them work well.

I see. I tested this on a real android device rather than the emulator so that i won't use AOSP devices as much as possible but I get a white blank image in 1.6.1 play store version of the FlowCrypt app instead of seeing the test image.

![Image](https://github.com/user-attachments/assets/de570819-0db3-41e0-8477-4e0189acc008)

> I can't reproduce it. If you still have this problem please make a video for me.

In my case, even the photo app or my google photo app is not getting suggested when tested in https://github.com/FlowCrypt/flowcrypt-android/pull/2948.

I'm not we'll aware of android development but is there a way to preview attachment like this.

Attachment preview is clicked by user -> initiates download -> invoked the file manager -> Open the attachment via file manager or some sort where we can imitate how file manager suggests apps with the "open with" .

![Image](https://github.com/user-attachments/assets/d95c7637-6184-4057-85b0-7caad341127b)
![Image](https://github.com/user-attachments/assets/94d4e0e0-349e-4326-944c-d61ca7e51c1d)

The reason i suggested this is that, I observed that file manager is very much versatile on this. In file manger, there's an option to open with selected app for capable apps of opening such file type. So this also answers fact, that microsoft word didn't remove the feature the open docx with it as I said on https://github.com/FlowCrypt/flowcrypt-android/issues/2944#issuecomment-2653125944 but rather its the FlowCrypt app's indeed needs improvement.

> I've improved this scenario. Please try.

This works nicely now. I can see prompt that the file already exists and no longer initiates automatic download. Thanks for this!

**Comment by martgil** [comment]
I'll be sharing the video screen recording for this one.

**Comment by DenBond7** [comment]
@martgil Thank you for the feedback!

> In my case, even the photo app or my google photo app is not getting suggested when tested in 
https://github.com/FlowCrypt/flowcrypt-android/pull/2948.

hm... interesting. Just in case, please share your device model(and manufacturer) and Android version. Also, will you be able to test the app via another real device?

> I'm not we'll aware of android development but is there a way to preview attachment like this.

>Attachment preview is clicked by user -> initiates download -> invoked the file manager -> Open the attachment via file manager or some sort where we can imitate how file manager suggests apps with the "open with" .

It's not possible. I can initiate opening a file via some app but I need to know some details about that app and this app should support such a behavior. This implementation will be app-oriented. Not good for other users.

> The reason i suggested this is that, I observed that file manager is very much versatile on this. In file manger, there's an option to open with selected app for capable apps of opening such file type. So this also answers fact, that microsoft word didn't remove the feature the open docx with it as I said on https://github.com/FlowCrypt/flowcrypt-android/issues/2944#issuecomment-2653125944 but rather its the FlowCrypt app's indeed needs improvement.

File managers work with files directly. It's a different story). Our implementation for previewing works differently. We care about security and don't store files locally. We load files in RAM and the receiver app can get access only via Uri access without the ability to change something.

To summarize the last conversation here and in #2944 please check the freshest changes and let me know about results.

And by the way, I have one more case how to check if all is going as expected. I will add details soon.

**Comment by DenBond7** [comment]
@martgil 
 
This one https://github.com/FlowCrypt/flowcrypt-android/pull/2948/commits/f7bbc4c40bd706b7a5b3336bee4260bf1972a72c - will force the app to show a long list of apps

This one https://github.com/FlowCrypt/flowcrypt-android/pull/2948/commits/59bbd7a4dcbdc65995349924e24e21445728cbe9 - will force the app to show a list of apps that support `<data android:mimeType="image/*" />`. Please try to preview an attachment(image). You should see Google Photo in the apps list. Also you will see a toast with the following text 
```
"Debug: Original = ${attachmentInfo.type}, final = ${Intent.normalizeMimeType(attachmentInfo.type)}"
```

Please share videos or screenshots for analyze.

**Comment by martgil** [comment]
> hm... interesting. Just in case, please share your device model(and manufacturer) and Android version. Also, will you be able to test the app via another real device?

My testing device (real android device) is iQOO Z9 Turbo running android 14. Unfortunately, I dont have another device to test aside from that and my emulator.

> This one https://github.com/FlowCrypt/flowcrypt-android/commit/f7bbc4c40bd706b7a5b3336bee4260bf1972a72c - will force the app to show a long list of apps

In my testing case, the MS Word for android or vivo document does not showed up with the list of compatible apps. Mime type from the toast is a truncated `application/vnd.openxmlformats-officedocument.wordprocessingml.document`.

> This one https://github.com/FlowCrypt/flowcrypt-android/commit/59bbd7a4dcbdc65995349924e24e21445728cbe9 - will force the app to show a list of apps that support <data android:mimeType="image/*" />. Please try to preview an attachment(image). You should see Google Photo in the apps list. Also you will see a toast with the following text

I have now seen google photo in the app list and able to view the photo from it very well. But I just wonder why the built in photo app is not available in app list.

> And by the way, I have one more case how to check if all is going as expected. I will add details soon.

Okay, just let me know if there's anything I could do to help you.
